File: native/android/core-model/src/main/kotlin/com/fireball/nativeapp/core/model/FireballModels.kt
package com.fireball.nativeapp.core.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class Track(
    val id: String,
    val videoId: String? = null,
    val title: String,
    val artist: String,
    val artwork: String? = null,
    val url: String? = null,
    val duration: Int? = null,
    val album: String? = null,
    val year: String? = null
) {
    val effectiveId: String get() = videoId ?: id
}

@Serializable
data class Playlist(
    val id: String,
    val title: String,
    val videos: List<Track> = emptyList()
)

@Serializable
data class Artist(
    val artistId: String,
    val name: String,
    val artwork: String? = null,
    val latestReleaseId: String? = null
)

@Serializable
data class Album(
    val id: String,
    val title: String,
    val artist: String,
    val artwork: String? = null,
    val year: Int? = null,
    val tracks: List<Track>? = null
)

@Serializable
data class FireballSettings(
    val ollamaEnabled: Boolean = false,
    val ollamaUrl: String = "",
    val ollamaModel: String = "llama3.2:3b",
    val lastFmApiKey: String = "",
    val lastFmApiSecret: String = "",
    val lastFmUsername: String = "",
    val lastFmSessionKey: String = "",
    val listenBrainzToken: String = "",
    val listenBrainzUsername: String = "",
    val highQuality: Boolean = false,
    val cacheEnabled: Boolean = true,
    /** In-memory search result cache cap (0 = [localMusicCacheLimit] × 10). */
    val searchCacheMaxEntries: Int = 0,
    /** Android Exo disk cache; iOS ignores. */
    val streamCacheEnabled: Boolean = true,
    val localMusicCacheLimit: Int = 10,
    val queueMode: String = "off",
    val invidiousInstance: String = "",
    val invidiousSid: String? = null,
    val invidiousUsername: String? = null,
    val sponsorBlock: Boolean = false,
    val sponsorBlockCategories: List<String> = emptyList(),
    val analytics: Boolean = false,
    val customDownloadPath: String? = null,
    val listenBrainzEnabled: Boolean = false,
    val listenBrainzPlayingNow: Boolean = false,
    val scrobbleEnabled: Boolean = true,
    val listenBrainzScrobblePercent: Int = ScrobbleRules.DEFAULT_PERCENT,
    val listenBrainzScrobbleMaxSeconds: Int = ScrobbleRules.DEFAULT_MAX_SECONDS,
    /** Tracks shorter than this are never scrobbled (Last.fm minimum; default 30s). */
    val listenBrainzScrobbleMinTrackSeconds: Int = ScrobbleRules.DEFAULT_MIN_TRACK_SECONDS,
    val invidiousPlaylistPrivacy: String = "private",
    val invidiousAutoPush: Boolean = false,
    val invidiousPlaylistMappings: Map<String, String> = emptyMap(),
    val webDavUrl: String = "",
    val webDavUsername: String = "",
    val webDavPassword: String = "",
    val gDriveEnabled: Boolean = false,
    val lastBackupAt: String? = null,
    val webDavLiveSync: Boolean = false,
    val remoteServerEnabled: Boolean = false,
    val remoteHostIp: String = "",
    val remotePeerPort: Int = 7771,
    val homeCountries: List<String> = emptyList(),
    val lyricsAutoScroll: Boolean = true,
    val lyricsReducedMotion: Boolean = false,
    val lyricsPreferEnglishHindi: Boolean = true,
    val themeMode: String = "system",
    val flexScheme: String = "deepPurple",
    val useDynamicColorWhenAvailable: Boolean = true,
    val accentSeedColor: Int? = null,
    val ipadSidebarCollapsed: Boolean = false,
    val gotifyEnabled: Boolean = false,
    val gotifyUrl: String = "",
    val gotifyToken: String = "",
    val lbdlUrl: String = "",
    val lbdlUsername: String = "",
    val lbdlPassword: String = "",
    val startTab: String = "home",
    val libraryUseGrid: Boolean = false,
    val offlineModeEnabled: Boolean = false,
    val privacyModeEnabled: Boolean = false,
    val dynamicIslandEnabled: Boolean = false,
    val loggingEnabled: Boolean = false,
    val bluetoothAutoplayEnabled: Boolean = false,
    val speakSongDetailsEnabled: Boolean = false,
    /** Android color source; `material_you` | `music` | `scheme`. Empty uses legacy [useDynamicColorWhenAvailable]. */
    val appearanceColorSource: String = "",
    /** Post local notifications when followed artists ship new albums (same probe as Gotify path). */
    val notifyArtistReleasesOnDevice: Boolean = false,
    /** Shows lyrics panel below transport on Now Playing (vs long-press on artwork default). */
    val alwaysShowLyricsPanel: Boolean = false,
)

@Serializable
data class PlaybackSession(
    val queue: List<Track> = emptyList(),
    val currentIndex: Int = 0,
    val shuffled: Boolean = false,
    val repeatMode: String = "off",
)

@Serializable
data class LibrarySnapshot(
    val version: Int = 2,
    val settings: FireballSettings = FireballSettings(),
    val history: List<Track> = emptyList(),
    val favorites: List<Track> = emptyList(),
    val playlists: List<Playlist> = emptyList(),
    val artists: List<Artist> = emptyList(),
    val albums: List<Album> = emptyList(),
    val playbackSession: PlaybackSession = PlaybackSession(),
)

File: native/ios/FireballNative/Core/FireballModels.swift
import Foundation

struct Track: Codable, Hashable {
    let id: String
    let videoId: String?
    let title: String
    let artist: String
    let artwork: String?
    let url: String?
    let duration: Int?
    let album: String?
    let year: String?

    var effectiveId: String { videoId ?? id }
}

struct Playlist: Codable, Hashable {
    let id: String
    let title: String
    let videos: [Track]
}

struct Artist: Codable, Hashable {
    let artistId: String
    let name: String
    let artwork: String?
    let latestReleaseId: String?
}

struct Album: Codable, Hashable {
    let id: String
    let title: String
    let artist: String
    let artwork: String?
    let year: Int?
    let tracks: [Track]?
}

struct FireballSettings: Codable, Hashable {
    var ollamaEnabled = false
    var ollamaUrl = ""
    var ollamaModel = "llama3.2:3b"
    var lastFmApiKey = ""
    var listenBrainzToken = ""
    var listenBrainzUsername = ""
    var highQuality = false
    var cacheEnabled = true
    var localMusicCacheLimit = 10
    var queueMode = "off"
    var invidiousInstance = ""
    var invidiousSid: String?
    var invidiousUsername: String?
    var sponsorBlock = false
    var sponsorBlockCategories: [String] = []
    var analytics = false
    var customDownloadPath: String?
    var listenBrainzEnabled = false
    var listenBrainzPlayingNow = false
    var listenBrainzScrobblePercent = 25
    var listenBrainzScrobbleMaxSeconds = 40
    var invidiousPlaylistPrivacy = "private"
    var invidiousAutoPush = false
    var invidiousPlaylistMappings: [String: String] = [:]
    var webDavUrl = ""
    var webDavUsername = ""
    var webDavPassword = ""
    var gDriveEnabled = false
    var lastBackupAt: String?
    var webDavLiveSync = false
    var remoteServerEnabled = false
    var remoteHostIp = ""
    var remotePeerPort = 7771
    var homeCountries: [String] = []
    var lyricsAutoScroll = true
    var lyricsReducedMotion = false
    var lyricsPreferEnglishHindi = true
    var themeMode = "system"
    var flexScheme = "deepPurple"
    var useDynamicColorWhenAvailable = true
    var accentSeedColor: Int?
    var ipadSidebarCollapsed = false
    var gotifyEnabled = false
    var gotifyUrl = ""
    var gotifyToken = ""
    var lbdlUrl = ""
    var lbdlUsername = ""
    var lbdlPassword = ""
    var offlineModeEnabled = false
    var privacyModeEnabled = false
    var dynamicIslandEnabled = false
    var loggingEnabled = false
    var bluetoothAutoplayEnabled = false
    var speakSongDetailsEnabled = false
    var startTab = "home"
    var libraryUseGrid = false
}

/// Restored on launch so queue / shuffle / repeat survive app restarts.
struct PlaybackSession: Codable, Hashable {
    var queue: [Track] = []
    var currentIndex: Int = 0
    var shuffled: Bool = false
    var repeatMode: String = "off"
}

struct LibrarySnapshot: Codable, Hashable {
    var version = 2
    var settings = FireballSettings()
    var history: [Track] = []
    var favorites: [Track] = []
    var playlists: [Playlist] = []
    var artists: [Artist] = []
    var albums: [Album] = []
    var playbackSession = PlaybackSession()
}
